cli/internal/runsummary/execution_summary.go

```diff
@@ -0,0 +1,201 @@
+	state     map[string]*TaskExecutionSummary `json:"-"`
+	Success   int                              `json:"success"`
+	Failure   int                              `json:"failed"`
+	Cached    int                              `json:"cached"`
+	Attempted int                              `json:"attempted"`
+
+	startedAt time.Time
+
+	profileFilename string
+}
+
+// newExecutionSummary creates a executionSummary instance to track events in a `turbo run`.`
+func newExecutionSummary(start time.Time, tracingProfile string) *executionSummary {
+	if tracingProfile != "" {
+		chrometracing.EnableTracing()
+	}
+
+	return &executionSummary{
+		Success:         0,
+		Failure:         0,
+		Cached:          0,
+		Attempted:       0,
+		state:           make(map[string]*TaskExecutionSummary),
+		startedAt:       start,
+		profileFilename: tracingProfile,
+	}
+}
+
+// Run starts the Execution of a single task. It returns a function that can
+// be used to update the state of a given taskID with the executionEventName enum
+func (es *executionSummary) run(label string) (func(outcome executionEventName, err error), *TaskExecutionSummary) {
+	start := time.Now()
+	taskExecutionSummary := es.add(&executionEvent{
+		Time:   start,
+		Label:  label,
+		Status: targetBuilding,
+	})
+
+	tracer := chrometracing.Event(label)
+
+	// This function can be called with an enum and an optional error to update
+	// the state of a given taskID.
+	tracerFn := func(outcome executionEventName, err error) {
+		defer tracer.Done()
+		now := time.Now()
+		result := &executionEvent{
+			Time:     now,
+			Duration: now.Sub(start),
+			Label:    label,
+			Status:   outcome,
+		}
+		if err != nil {
+			result.Err = fmt.Errorf("running %v failed: %w", label, err)
+		}
+		// Ignore the return value here
+		es.add(result)
+	}
+
+	return tracerFn, taskExecutionSummary
+}
+
+func (es *executionSummary) add(event *executionEvent) *TaskExecutionSummary {
+	es.mu.Lock()
+	defer es.mu.Unlock()
+	if s, ok := es.state[event.Label]; ok {
+		s.Status = event.Status.toString()
+		s.Err = event.Err
+		s.Duration = event.Duration
+	} else {
+		es.state[event.Label] = &TaskExecutionSummary{
+			StartAt:  event.Time,
+			Label:    event.Label,
+			Status:   event.Status.toString(),
+			Err:      event.Err,
+			Duration: event.Duration,
+		}
+	}
+	switch {
+	case event.Status == TargetBuildFailed:
+		es.Failure++
+		es.Attempted++
+	case event.Status == TargetCached:
+		es.Cached++
+		es.Attempted++
+	case event.Status == TargetBuilt:
+		es.Success++
+		es.Attempted++
+	}
+
+	return es.state[event.Label]
+}
+
+// writeChromeTracing writes to a profile name if the `--profile` flag was passed to turbo run
+func writeChrometracing(filename string, terminal cli.Ui) error {
+	outputPath := chrometracing.Path()
+	if outputPath == "" {
+		// tracing wasn't enabled
+		return nil
+	}
+
+	name := fmt.Sprintf("turbo-%s.trace", time.Now().Format(time.RFC3339))
+	if filename != "" {
+		name = filename
+	}
```

Comment:
Saves some work

```suggestion
	if filename == "" {
		filename = fmt.Sprintf("turbo-%s.trace", time.Now().Format(time.RFC3339))
	}
```

Reply:
i generally prefer not to reassign incoming function args, but in this case I also don't really want to make changes here since this function is not the focus of this PR (it just got moved around)